src/Features/CSharp/Portable/UsePatternMatching/CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer.cs

```diff
@@ -179,12 +179,12 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext context)
         }
 
         public SemanticModel ReplaceMatches(
-            Workspace workspace, SemanticModel semanticModel, BinaryExpressionSyntax isExpression,
+            SemanticModel semanticModel, BinaryExpressionSyntax isExpression,
             string localName, HashSet<CastExpressionSyntax> matches,
             CancellationToken cancellationToken)
         {
             var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
-            var editor = new SyntaxEditor(root, workspace);
+            var editor = new SyntaxEditor(root, CSharpSyntaxGenerator.Instance);
```

Comment:
`SyntaxEditor` is still a workspace dependency and it will prevent this analyzer from moving to shared layer. Ideally, analyzers shouldn't be doing any "Replace" operations or forking semantic model to check if the replacement causes errors. We should instead try to use speculative semantic model or `SpeculationAnalyzer` to identify if replacement causes errors.

Reply:
I believe it's similar to #36264

Reply:
Agreed, this change by itself is good, but you would need a similar change to the analyzer.

Reply:
Also note that the only public constructor of SyntaxEditor takes a Workspace parameter, so you would not be able to execute this analyzer in CI with SyntaxEditor usage.

Reply:
> We should instead try to use speculative semantic model or SpeculationAnalyzer to identify if replacement causes errors.

Speculative models don't support data flow analysis.  I talked with neal about this, and this seems to be the best way currenlty (modulo reimplementing data flow on the ide side) to know if this sort of change is ok.

Reply:
> SyntaxEditor is still a workspace dependency 

That's super weird.  All it is is annotations and helper methods that sit on top of the compiler API.  Can we not move some of it down?  The only reason it has any Workspace today is just for discovery of the SyntaxGenrator.  The actual state of it is just:

```
        private readonly SyntaxGenerator _generator;
        private readonly List<Change> _changes;
        private bool _allowEditsOnLazilyCreatedTrackedNewNodes;
        private HashSet<SyntaxNode> _lazyTrackedNewNodesOpt;
```

Reply:
Yep, it is a bigger work item to move SyntaxGenerator down. Especially with the fact that it uses bunch of public annotations from Workspaces layer (simplifier, formatter, etc. annotations). We cannot just clone these annotations into shared layer because I think the simplifier and formatter will not recognize these. Moving anything that relies on syntax changes + adding annotations into the shared analyzer layer is going to be tricky.

Reply:
> We cannot just clone these annotations into shared layer because I think the simplifier and formatter will not recognize these.

We actually can. Annotations are done with a simple string-key.  So as long as use that key, we're fine :)